File: 0104. Maximum Depth of Binary Tree/solution_1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
    public:
        int maxDepth(TreeNode* root) 
        {
            int depth = 0;

            if (root)
            {
                vector<TreeNode*> nodes = {root};

                while (!nodes.empty())
                {
                    vector<TreeNode*> nextNodes;

                    for (auto node : nodes)
                    {
                        if (node->left)
                            nextNodes.push_back(node->left);
                        
                        if (node->right)
                            nextNodes.push_back(node->right);
                    }

                    nodes.clear();
                    nodes.assign(nextNodes.begin(), nextNodes.end());

                    depth++;
                }

            }

            return depth;
        }
};

File: 0104. Maximum Depth of Binary Tree/solution_1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

 class Solution
 {
     public int maxDepth(TreeNode root)
     {
         int depth = 0;
 
         if (root != null)
         {
             ArrayList<TreeNode> nodes = new ArrayList<>();
             nodes.add(root);
 
             while (!nodes.isEmpty())
             {
                 ArrayList<TreeNode> nextNodes = new ArrayList<>();
                 
                 for (var node : nodes)
                 {
                     if (node.left != null)
                         nextNodes.add(node.left);
                     
                     if (node.right != null)
                         nextNodes.add(node.right);
                 }
                 
                 nodes.clear();
                 nodes.addAll(nextNodes);
                 
                 depth++;
             }
         }
         
         return depth;
     }
 }